light_curves/code_src/icecube_functions.py

```diff
@@ -44,67 +39,52 @@ def icecube_get_lightcurve(coords_list, labels_list, icecube_select_topN=3, verb
-            icecube_matches[index]["Ang_match"] = dist_angle.to(u.degree).value[sel[0:this_topN]]
-            icecube_matches[index]["Ang_match"].unit = u.degree
-            icecube_matched.append(objectid)
-
-        else:
-            if verbose:
-                print("No match found.")
-
-    # Add to lightcurve object
-    for index, (objectid, coord) in enumerate(tqdm(coords_list)):
-        label = labels_list[index]
-        if objectid in icecube_matched:
-            # Create single instance. flux in log GeV, error is N/A & set to 0, time is MJD
-            dfsingle = pd.DataFrame({'flux': np.asarray(icecube_matches[index]["energy_logGeV"]),
-                                     'err': np.repeat(0, len(icecube_matches[index])),
-                                     'time': np.asarray(icecube_matches[index]["mjd"]),
-                                     'objectid': np.repeat(objectid, len(icecube_matches[index])),
-                                     'label': label,
-                                     'band': "IceCube"}).set_index(["objectid", "label", "band", "time"])
-
-            df_lc.append(dfsingle)
-
-    if verbose:
-        print("IceCube Matched and added to lightcurve object.")
-
-    return (df_lc)
+    # create SkyCoord objects from icecube event coordinates
+    icecube_skycoords = SkyCoord(icecube_events["ra"], icecube_events["dec"], unit="deg", frame='icrs')
+
+    #here are the skycoords from mysample defined above
+    mysample_skycoords = sample_table['coord']
+
+    #Match
+    idx_icecube, idx_mysample, d2d, d3d = icecube_skycoords.search_around_sky(mysample_skycoords, 1*u.deg)
+
+    #need to filter reponse based on position error circles
+    #really want d2d to be less than the error circle of icecube = icecube_events["AngErr"] in degrees
+    filter_arr = d2d < icecube_events["AngErr"][idx_mysample]
+    filter_idx_icecube = idx_icecube[filter_arr]
+    filter_idx_mysample = idx_mysample[filter_arr]
+    filter_d2d = d2d[filter_arr]
+
+    #keep these matches together with objectid and lebal as new entries in the df.
+    obid_match = sample_table['objectid'][filter_idx_icecube]
+    label_match = sample_table['label'][filter_idx_icecube]
+    time_icecube= icecube_events['mjd'][filter_idx_mysample]
+    flux_icecube = icecube_events['energy_logGeV'][filter_idx_mysample]
+
+    #save the icecube info in correct format for the rest of the data
+    icecube_df = pd.DataFrame({'flux': flux_icecube, 
+                               'err': np.zeros(len(obid_match)), 
+                               'time': time_icecube, 
+                               'objectid': obid_match, 
+                               'label': label_match, 
+                               'band': "IceCube"})
+
+    #now can use a groupby to only keep the top N (by GeV flux) icecube matches for each object
+    filter_icecube_df = icecube_df.groupby('objectid').head(icecube_select_topN).reset_index(drop=True)
```

Comment:
Currently this is working correctly and keeping only the largest fluxes (at least for the Yang sample), but all the indexing and filtering that happen between the sort (line 49) and this line is dangerous. In particular, the astropy docs say that the index arrays returned by `search_around_sky` may be sorted differently in the future, which could break this code.

I recommend doing the sort immediately prior to this line. Something like:

```python
icecube_df = icecube_df.sort_values(['objectid', 'flux'], ascending=[True, False])
```

Then I think you can delete line 49.

Reply:
thanks, I didn't know about `sort_values`, this was helpful.